### ISSUE TITLE
Validation error: "no labels found in val set, can not compute metrics without labels"

### ISSUE THREAD

**Issue by GxTeo** [ultralytics/yolov5#8791]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have checked my annotations and train & validation sets. Everything seems to be fine but I have been getting this warning whenever I want to train my model 
![image](https://user-images.githubusercontent.com/81504451/181880278-bb2008eb-568d-4122-ae16-e1860ab8eed0.png)

**Comment by glenn-jocher** [comment]
@GxTeo 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. Your dataset has no validation labels or is incorrectly formatted an no validation labels can be found.

Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by FranciosoDonato** [comment]
Hi, i have the same error. This morning no problem, but in this afternoon i found this warning and i have't change nothing. I use RoboFlow for create my dataset and i export it in my code. 
`from roboflow import Roboflow`
`rf = Roboflow(api_key="<key>")`
`project = rf.workspace("computer-vision-ppbyu").project("rice-detection-mnpf6")`
`dataset = project.version(30).download("yolov5")`
`!python train.py --batch 64 --epochs 20 --data '/content/yolov5/dataset/data.yaml' --weights yolov5s.pt`

I follow all the tutorial but the warning persists. Cause this error my metric are not visualized correctly.

**Comment by glenn-jocher** [comment]
@FranciosoDonato got it. You might want to raise the issue with the Roboflow team, there may be a dataset issue as the command works correctly with the default datasets, i.e. 
```python
!python train.py --batch 64 --epochs 20 --data coco128.yaml --weights yolov5s.pt
```

**Comment by GxTeo** [comment]
I have tried the same dataset with no changes to the labels folder using google colab and had no error so its unlikely an annotation format problem. Any other possible reasons why I might have such errors?

**Comment by glenn-jocher** [comment]
@GxTeo I'll take another look at this. This message was recently introduced in https://github.com/ultralytics/yolov5/pull/8782

**Comment by glenn-jocher** [comment]
@GxTeo yes it looks like the labels count is not considered if no detections are found, i.e.:

```
 python val.py --imgsz 32 --conf 0.001
 python val.py --imgsz 32 --conf 0.999
```

I'll work on a fix.

**Comment by glenn-jocher** [comment]
@GxTeo good news 😃! Your original issue may now be fixed ✅ in PR #8806. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by GxTeo** [comment]
@glenn-jocher Hi, thank you so much! It works perfectly now!

**Comment by RonaldYuren** [comment]
Hi I have some question about the similar question.
![image](https://user-images.githubusercontent.com/95292013/184881780-8efca4ef-9912-4878-a821-e87b2fc58f9c.png)





I didn't find the place to add labels' path, I am trying to run the yolov7 custom_dataset
but the train.txt and valid.txt I only put my images' path；
Are there any place I need to add my own labels?

**Comment by GxTeo** [comment]
@RonaldYuren In Yolov5, i am sure the data yaml source file directly indicates the path to your images. There is no train.txt or val.txt file needed for the paths of images which is used in Yolov4 and Yolov7. Perhaps you can screenshot and share your yaml source file here?

**Comment by RonaldYuren** [comment]
> 

I have pass the the question about the path, I am wondering that what the detect.py will do and how to use detect.py ?
I think that detect.py is similar for v5 & v7 right??

**Comment by glenn-jocher** [comment]
@RonaldYuren detect.py Usage examples are shown right in the file:
https://github.com/ultralytics/yolov5/blob/6728dad76df8d62ed3c08e39c224a773d20582a0/detect.py#L1-L26

**Comment by Ahmadreza1365** [comment]
I have similar issue with yolov7.  I read all the comments above, but I couldn't resolve it.

![1](https://user-images.githubusercontent.com/89111238/204978952-7ffa1450-a66d-42ef-ab20-a5d31db8ebce.jpg)

**Comment by Linkhdaiong** [comment]
> * https://github.com/ultralytics/yolov5



> ### 询问前搜索
> * [x] 我搜索了 YOLOv5 问题和[讨论](https://github.com/ultralytics/yolov5/discussions)，没有发现类似的[问题](https://github.com/ultralytics/yolov5/issues)。
> 
> ### 问题
> 我已经检查了我的注释和训练与验证集。一切似乎都很好，但是每当我想训练我的模型时，我都会收到此警告 ![image](https://user-images.githubusercontent.com/81504451/181880278-bb2008eb-568d-4122-ae16-e1860ab8eed0.png)
> 
> ### 附加
> _无响应_
您好！您是怎么解决的呢，我的问题跟你一样，也重新下载了源码，还是没有val.label

**Comment by glenn-jocher** [comment]
@Linkhdaiong 您好！感谢您的提问！对于您遇到的问题，可能是由于在验证集中没有找到标签导致的。

警告消息是提示没有在验证集中找到标签，因此无法计算指标。在使用YOLOv5时，验证集需要包含有标签的数据，以便进行精度评估和模型性能分析。

请确保您的验证集中的每个图像都有对应的标签文件，并且标签文件的格式正确。标签文件通常是以相同名称但不同扩展名（如.jpg和.txt）命名的文件，其中.txt文件包含了相应图像的目标物体的边界框和类别信息。

如果您确认标签文件存在并且格式正确，但仍然出现此问题，请检查文件的路径和命名是否正确。确保将数据集的标签文件路径正确指定到代码中。

希望这些提示能对您有所帮助！如果您有任何其他问题或者需要进一步的帮助，请随时告诉我们。我们很乐意帮助您解决问题！

**Comment by Miarviolin** [comment]
hi , I have the same error 
![image](https://github.com/ultralytics/yolov5/assets/13646162/f820a7bb-def7-4446-ab18-c6ecd214d408)
and this how my files look 
![image](https://github.com/ultralytics/yolov5/assets/13646162/c0180790-b25b-49f8-8a3b-9a32c088046b)
thanks inadvance

**Comment by MeatyAri** [comment]
For those encountering the same error, like me and @Linkhdaiong, that displays the message:
`WARNING no labels found in detect set, can not compute metrics without labels`

Here's a solution: Organize your label files in a structure similar to your image directories. For instance, if you have a train/validation split like me:
![image](https://github.com/ultralytics/yolov5/assets/68596698/e1ac4585-d41d-4919-8554-7c1d5b75c26d)
Make sure to create corresponding train/validation split directories for your labels as well.

A helpful tip: You can simply copy all the label files to both directories. YOLO will use only the ones it needs, so there's no need to worry about duplicating files.

**Comment by glenn-jocher** [comment]
Hi @MeatyAri, thank you for bringing this issue to our attention. I understand that you and @Linkhdaiong are encountering the same error message, "WARNING no labels found in detect set, can not compute metrics without labels." 

To resolve this issue, please ensure that your label files are organized in a structure similar to your image directories. This means creating corresponding train/validation split directories for your labels. You can refer to the provided image for an example structure.

A useful tip is to copy all the label files to both directories. YOLOv5 will only use the necessary label files, so there is no need to worry about duplicating files.

I hope this solution helps resolve the issue for you. Please let us know if you need any further assistance.

**Comment by Andres-San** [comment]
Hi There, 

After "loosing" a few hours, I figured that it was not a path issue, moreover it was a format issue: 

- Format of my label files, was not correct. I had not the format "class x_center  y_center width height" as specified [here](https://docs.ultralytics.com/datasets/detect/#ultralytics-yolo-format) 

- If you are using google Collab, when uploadind to .txt files to GDrive, they are converted to .gdocs, so you must prevent this from happening (settings in drive)

When I resolved both those issues, it run well for me. 

For anyone wondering, I labeled my photos using [labelme](https://github.com/labelmeai/labelme), and then used [labelme2yolo](https://github.com/greatv/labelme2yolo#readme) to get the yolo format. Problem was that labelme2yolo has the default output format to polygon, so I changed it to bbox and problem solved regarding the yolo format. 

I guess the warning was **pretty missleading**, instead for "labels not found" it should be something more like "wrong format" 

Hope it helps somebody, Best!

**P.S:** Before all of this, make sure you have your paths right (no labels path required in your .yaml, only images, and label path will mimic the image path ) 
**P.S2:** My final clue was that in the error message, at some point it says _"no label found in :{PATH}"_ and that PATH was exactly the correct Path, so then I started digging.

**Comment by glenn-jocher** [comment]
@Andres-San hi There,

Thank you so much for sharing your findings and solutions! 🙌 It's super helpful for the community.

To summarize and add a bit:
- Always ensure label files follow the "class x_center y_center width height" format as per Ultralytics Yolo format guidelines.
- When using Google Colab and Google Drive, be mindful of file conversions (like .txt to .gdocs). Adjusting Drive settings can prevent this.
- Tools like labelme are great, but converting formats (e.g., with labelme2yolo) requires careful attention to details like output folders. Switching from 'polygon' to 'bbox' for YOLO format is a key step.
- A "wrong format" error indicator would indeed be more precise than a general "labels not found" message. Your suggestion is noted for improvement! 📝
- Lastly, confirming the path is crucial, as the correct path can still result in errors if the format or file standards aren't met.

Your PS notes perfectly highlight the importance of verifying paths and the format. Your journey and solution will definitely assist others facing similar issues. 

Best regards and happy coding!

**Comment by Bakhtawar378** [comment]
i was facing the same problem and it turned out that my "images" folder was "Images" and labels folder was "Labels"... i just changed the case for I and L and the problem was solved

**Comment by glenn-jocher** [comment]
@Bakhtawar378 hi there! 😊

Great to hear you've resolved the issue! Indeed, the case sensitivity in file or folder names can lead to such problems. It's always good to double-check the file & folder names match exactly what the code expects. Happy coding, and feel free to reach out if you encounter any more issues. Keep up the great work! 👍
